### ISSUE TITLE
Synchronize asynchronous STDIO streams

### ISSUE THREAD

**Issue by 6-8-axnw1bom81v5xa3nh48c** [nodejs/help#921]
How to synchronously read data from `stdout` of one process and synchronously write to another process's `stdin`?

### The problem

I have an application which spawns two processes **A** and **B** (using `child_process`). Then I want to have a loop where I read **A**'s data, process it and write to **B**. The problem is because my program may be faster than **A**'s required time to generate data and **B**'s time to get ready for receiving new data. Also, my program may be slower than **A** and **B** too. So, I need to somehow synchronize them.

### My solution

The solution is simple. Attach `data` event listener to **A**'s `stdout`, then process data and store them somewhere until **B** is ready.  The problem again is because I want to read from **A** independently of **B** and I want to be able to write to **B** even if **A** didn't send any data yet. Basically, what I want is to **synchronously** write to **A** and **B**. I'm looking for the correct syntax to achieve it.

### Example

Here is an example of how I would like to write my code (integer function arguments represent number of bytes):

```JavaScript
var input = A.read(50);

for(i = 0; i < someVariable; i++){
  var buff = A.read(5);
  if(someCheck(buff)){
    B.write(buff);
  }else{
    B.write(input);
  }
}

B.write(A.read(buff[0]));
```

This example is completely fictional. I just tried to explain what I'm trying to do. So, basically, I want to synchronously read and write from/to **A** and **B** at any time (even in a loop). The `read` method reads `n` bytes and returns corresponding buffer, and `write` writes that buffer.

### Idea

What I'm trying to do is definitely possible, but I'm just asking for the easiest way to achieve it. This is my idea (but I'm not sure how to exactly implement it and would it be efficient enough):

1. Create a new module which will be able to spawn a new thread
2. Create another module which will asynchronously receive data from spawned child process
3. Implement `read` like this: create new thread, spin in a loop until another thread asynchronously reads enough bytes, then break the loop and construct a buffer
4. Implement `write` similarly.

The problem with this approach is because spinning in a loop is very bad practice, it uses more than 50% of CPU and therefore unnecessarily slows down **A**'s and **B**'s working speed. How to pause the thread until enough bytes are received by another thread?

Any ideas on how to implement this, or is there any better way? Thanks anyone who decide to help :)

**Comment by bnoordhuis** [comment]
Synchronously?  No, that's antithetical to Node.js, but don't streams work for you?  You can read from and write to A and B independently.

**Comment by 6-8-axnw1bom81v5xa3nh48c** [comment]
@bnoordhuis Thank you for your comment. I know synchronizing streams is not something Node developers will recommend, but I'm still trying to do it anyway. It is definitely possible, I've already done it using multithreading module and spinning in a loop until enough data gets read by another thread. I've also explained why is spinning in a loop bad approach, so I'm searching for another way.

*"You can read from and write to A and B independently."* - Either I don't understand what do you want to say, or maybe you didn't get from my issue (question) what I'm trying to achieve. Of course, it is possible to read/write independently, but can I read from A, process the obtained buffer and write the processed buffer to B synchronously in a loop? I don't think native Node.js syntax allows it.

Thank you anyway for your fast response. Please don't get me wrong, but I'll prefer to leave this question open until someone else come with a good idea.

**Comment by bnoordhuis** [comment]
> can I read from A, process the obtained buffer and write the processed buffer to B synchronously in a loop?

Yes, apart from the word "synchronously".

Why is it important that is synchronous?  What is your definition of "synchronous"?

**Comment by 6-8-axnw1bom81v5xa3nh48c** [comment]
@bnoordhuis
Sir, can you please take a look at the code snippet from my question? I hope it clearly explains what I'm trying to achieve.

*"Why is it important that is synchronous?"* - I already have a code which expects these methods to be synchronous. The code is about ~10K lines of code, so I definitelly cannot asynchronize and restructure whole code, I need to implement synchronously `read` and `write` methods.

*"What is your definition of "synchronous"?"* Application starts, the main thread enters a loop, read from A, write to B (in a loop), when there is no more data inside A, then break the loop and the main thread exits. Very simple, no event emitters, event listeners, `data` event handlers, attaching listeners to stdout, nothing. Simply function calls and loop and that's it.

Node.js provides blocking `fs.readSync` and `fs.writeSync` methods, but there is nothing similar for `stdin` and `stdout`, you need to attach event listener for `data` event and wait until new data is received. But, I'm not looking for a native Node.js way to achieve blocking stdio, I'm looking for methods I've explained in my question (for example multithreading: the first thread synchronously reads from the second thread, while the second thread asynchronously waits for `data` callback).

Thank you anyway for trying to help. I appreciate it.

**Comment by bnoordhuis** [comment]
> I hope it clearly explains what I'm trying to achieve.

It does but my point was that it can be rewritten to use streams (i.e., async I/O) with no change in observable behavior.  However...

> I already have a code which expects these methods to be synchronous. The code is about ~10K lines of code, so I definitelly cannot asynchronize and restructure whole code

...that means you're probably out of luck.  Node.js has no built-in support for what you want to do.

A Linux-specific hack that might work is this:
```js
var afile = '/proc/self/fd/' + A.stdout._handle.fd;
var bfile = '/proc/self/fd/' + B.stdin._handle.fd;

var afd = fs.openSync(afile, 'r');  // reopen
var bfd = fs.openSync(bfile, 'w');  // reopen

// now read from |afd| and write to |bfd|
var buf = Buffer.alloc(5);
var num = fs.readSync(afd, buf, 0, buf.length, -1);
fs.writeSync(bfd, buf, 0, num, -1);
```
Untested, though.  Use at your own risk.

**Comment by 6-8-axnw1bom81v5xa3nh48c** [comment]
@bnoordhuis
I knew about that hack, but unfortunately I'm on Windows.
I decided to try multithreading approach (because as it seems Node doesn't have a native way to do what I want). My approach works, but the only problem is Node's main thread (which spins in a loop until the second thread asynchronously reads the data) eats CPU. It uses more than 50% of CPU. Is there a way to calm it down?

Basically, what I currently have in the main thread is

```JavaScript
A.read = size => {
  global.myBuffer = Buffer.from([]);
  while(global.myBuffer.length != size);
  return global.myBuffer;
};
```

This code resets the buffer and waits until the second thread asynchronously read enough bytes and then returns that buffer. However, the line `while(global.myBuffer.length != size);` takes about 52% of CPU. Which techniques Node.js uses to prevent huge CPU usage in native event listeners? For example, the following code:

```JavaScript
var t = Date.now();
while(Date.now() - t < 10e3);
console.log('ok');
```

eats CPU, while this one doesn't:

```JavaScript
setTimeout(() => console.log('ok'), 10e3);
```

What is the difference? How does Node succeed to silence the thread until the event listener is ready to be called? To avoid confusion, I know Node doesn't use multiple threads, it has one thread, but I'm spawning another thread using a module for that purpose. But, my question is related to something else: how does Node wait for callbacks so silently? I mean, somewhere in C++ land thread spins in a loop until some external data change, right? How then it doesn't eat CPU like my `while` loop?

Thank you for your responses and sorry if my questions are too annoying.

**Comment by ORESoftware** [comment]
intercept the stream, `unpipe` it from everything, then pipe it to an `on('data')` handler that uses `fs.writeSync`?

**Comment by bnoordhuis** [comment]
> Which techniques Node.js uses to prevent huge CPU usage in native event listeners?

It suspends itself by entering a system call like epoll_wait() or GetQueuedCompletionStatusEx().

In order for that to work, control needs to return to the event loop.  That doesn't happen if you have a busy loop.

You could ameliorate CPU usage by adding a condition variable to your add-on that JS code can wait on or simply by calling `Sleep(1)` repeatedly.

**Comment by 6-8-axnw1bom81v5xa3nh48c** [comment]
`GetQueuedCompletionStatusEx` is exactly what I was looking for.

Thanks!
